src/controllers/auth.controller.ts

```diff
@@ -77,4 +77,59 @@ const login = async (req: Request, res: Response) => {
   }
 };
 
-export { register, login };
+const oAuth2Client = new OAuth2Client(
+  process.env.GOOGLE_CLIENT_ID,
+  process.env.GOOGLE_CLIENT_SECRET,
+  'postmessage',
+);
+
+const loginWithGoogle = async (req: Request, res: Response) => {
+  const {code} = req.body;
+  if (!code) {
+    return res.status(400).json({
+      success: false,
+      message: 'Missing information.',
+    });
+  }
+  try {
+    const { tokens } = await oAuth2Client.getToken(code);
+    //get accessToken and expire
+    const accessToken = tokens.access_token;
+    const expiresIn = tokens.expiry_date;
+    //get username = id, email
+    const decoded = jwt.decode(tokens.id_token) as jwt.JwtPayload;
+    const username = decoded.sub;
+    const email = decoded.email;
+    //check if email is taken
+    const isEmailTaken = await checkEmail({username, email});
+    if (isEmailTaken) {
+      return res.status(409).json({
+        success: false,
+        message: 'Email is already taken',
+      });
+    }
+    const user = await findUser({ username , email });
+    if (user) {
+      return res.json({
+        success: true,
+        message: 'Log in successfully again.',
+        accessToken,
+        expiresIn,
+      });
+    }
+    await createUser(decoded.name, email, "0912345678", username as string, username as string);
```

Comment:
What is `0912345678` ?

Reply:
Google OAuth does not return phone number so it's just a random number

Reply:
Can we set it `null`?

Reply:
Yes